libs/types/src/tabulation.ts

```diff
@@ -0,0 +1,130 @@
+  readonly contestType: 'candidate';
+  readonly votesAllowed: number;
+  readonly tallies: Record<CandidateId, CandidateTally>;
+};
+
+/**
+ * Represents the results of a single contest in an election, often filtered by
+ * some cast vote record attributes.
+ */
+export type ContestResults = YesNoContestResults | CandidateContestResults;
+
+/**
+ * Should match `CVR.vxBallotStyle`.
+ */
+export type VotingMethod = 'absentee' | 'precinct' | 'provisional';
+
+/**
+ * Indicates what cast vote records to include when calculating results.
+ * Omission of a filter attribute indicates *not* filtering on it at all.
+ * So an empty `Filter` of `{}` would indicate including all cast
+ * vote records in the results.
+ */
+export interface Filter {
+  readonly ballotStyleIds?: BallotStyleId[];
+  readonly partyIds?: Id[];
+  readonly precinctIds?: PrecinctId[];
+  readonly votingMethods?: VotingMethod[];
+  readonly batchIds?: Id[];
+  readonly scannerIds?: Id[];
+}
+
+/**
+ * Attributes that always exist for every cast vote record.
+ */
+export interface CastVoteRecordAttributes {
+  readonly ballotStyleId: BallotStyleId;
+  readonly precinctId: PrecinctId;
+  readonly votingMethod: VotingMethod;
+  readonly batchId: Id;
+  readonly scannerId: Id;
+}
+
+/**
+ * Scanned cards can either by BMD ballots or a sheet of an HMPB ballot,
+ * indicated by its 1-indexed number.
+ */
+export type CardType = 'bmd' | number;
+
+/**
+ * In situations where we're generating grouped results, specifiers can be
+ * included in {@link ElectionResult} to indicate what it is a grouping of.
+ */
+export type GroupSpecifier = Partial<CastVoteRecordAttributes> & {
+  readonly partyId?: Id;
+};
+
+/**
+ * The cast vote record attributes we can use to group election results. For
+ * example, you may have the full election results grouped by ballot style.
+ * The grouping options are all based on {@link CastVoteRecordAttributes}
+ * except for `party`, which is determined based on `ballotStyle`.
+ */
+export interface GroupBy {
+  groupByBallotStyle?: boolean;
+  groupByParty?: boolean;
+  groupByPrecinct?: boolean;
+  groupByVotingMethod?: boolean;
+  groupByBatch?: boolean;
+  groupByScanner?: boolean;
+}
+
+export type CardCounts = Record<CardType, number>;
+
+/**
+ * Represents the results of all contests in an election, often filtered by
+ * some cast vote record attributes.
+ */
+export interface ElectionResults {
+  readonly contestResults: Record<ContestId, ContestResults>;
+  readonly cardCounts: CardCounts;
+}
+
+export type GroupKey = string;
+export type GroupOf<T> = T & GroupSpecifier;
+export type Grouped<T> = Record<GroupKey, GroupOf<T>>;
+
+export type GroupedElectionResults = Grouped<ElectionResults>;
+
+/**
+ * Simplified representation of votes on a scanned ballot for tabulation
+ * purposes.
+ */
+export type Votes = Record<ContestId, Id[]>;
+
+export type CastVoteRecord = {
+  readonly votes: Votes;
+  readonly cardType: CardType;
```

Comment:
I think that it would make sense to have it more explicit like you've typed out. 

Relatedly, I'm still not sure how to represent the `CardCounts` type for `ElectionResults`. The reason I had it combined before was so I could just use the `CardType` as an index in the `CardCounts`. But that has other problems - using a `number` as an index is not ideal, because it's converted to a string. Still thinking about how to make `CardCounts`.